objectivec/include/ort_training_session.h

```diff
@@ -0,0 +1,261 @@
+ * @param trainModelPath The path to the training onnx model.
+ * @param evalModelPath The path to the evaluation onnx model.
+ * @param optimizerModelPath The path to the optimizer onnx model used to perform gradient descent.
+ * @param error Optional error information set if an error occurs.
+ * @return The instance, or nil if an error occurs.
+ *
+ */
+- (nullable instancetype)initWithEnv:(ORTEnv*)env
+                      sessionOptions:(ORTSessionOptions*)sessionOptions
+                          checkPoint:(ORTCheckpoint*)checkPoint
+                      trainModelPath:(NSString*)trainModelPath
+                       evalModelPath:(nullable NSString*)evalModelPath
+                  optimizerModelPath:(nullable NSString*)optimizerModelPath
+                               error:(NSError**)error NS_DESIGNATED_INITIALIZER;
+
+/**
+ * Performs a training step, which is equivalent to a forward and backward propagation in a single step.
+ *
+ * The training step computes the outputs of the training model and the gradients of the trainable parameters
+ * for the given input values. The train step is performed based on the training model that was provided to the training session.
+ * It is equivalent to running forward and backward propagation in a single step. The computed gradients are stored inside
+ * the training session state so they can be later consumed by `optimizerStep`. The gradients can be lazily reset by
+ * calling `lazyResetGrad` method.
+ *
+ * @param inputs The input values to the training model.
+ * @param error Optional error information set if an error occurs.
+ * @return The output values of the training model.
+ */
+- (nullable NSArray<ORTValue*>*)trainStepWithInputValues:(NSArray<ORTValue*>*)inputs
+                                                   error:(NSError**)error;
+
+/**
+ * Performs a evaluation step that computes the outputs of the evaluation model for the given inputs.
+ * The eval step is performed based on the evaluation model that was provided to the training session.
+ *
+ * @param inputs The input values to the eval model.
+ * @param error Optional error information set if an error occurs.
+ * @return The output values of the eval model.
+ *
+ */
+- (nullable NSArray<ORTValue*>*)evalStepWithInputValues:(NSArray<ORTValue*>*)inputs
+                                                  error:(NSError**)error;
+
+/**
+ * Reset the gradients of all trainable parameters to zero lazily.
+ *
+ * Calling this method sets the internal state of the training session such that the gradients of the trainable parameters
+ * in the ORTCheckpoint will be scheduled to be reset just before the new gradients are computed on the next
+ * invocation of the `trainStep` method.
+ *
+ * @param error Optional error information set if an error occurs.
+ * @return YES if the gradients are set to reset successfully, NO otherwise.
+ */
+- (BOOL)lazyResetGradWithError:(NSError**)error;
```

Comment:
Is there a reason we need to return a bool considering we have `error` to indicate success/failure?

Reply:
Maybe this is commonplace in objective-c and is the right thing to do.

Reply:
I believe it is standard practice for Objective-C methods to return `BOOL` indicating success. I believe we also need to return boolean to support translation to [swift](https://github.com/apple/swift/blob/main/docs/CToSwiftNameTranslation.md#error-handling).